dbt_style_guide.md

```diff
@@ -1,237 +1,508 @@
+
+- CTE names should be as verbose as needed to convey what they do.
+
+- CTEs with confusing or noteable logic should be commented with SQL comments as you would with any complex functions, and should be located above the CTE.
+
+- CTEs that are duplicated across models should be pulled out and created as their own models.
+
+- CTEs fall in to two main categories:
+  | Term    | Definition                                                                                                                                                             |
+  |---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+  | Import  | Used to bring data into a model. These are kept relatively simple and refrain from complex operations such as joins and column transformations.                        |
+  | Logical | Used to perform a logical step with the data that is brought into the model toward the end result. |
+
+- All `{{ ref() }}` or `{{ source() }}` statements should be placed within import CTEs so that dependent model references are easily seen and located.
+
+- Where applicable, opt for filtering within import CTEs over filtering within logical CTEs. This allows a developer to easily see which data contributes to the end result.
+
+- SQL should end with a simple select statement. All other logic should be contained within CTEs to make stepping through logic easier while troubleshooting.
+  Example: `select * from final`
+
+- SQL and CTEs within a model should follow this structure:
+  - `with` statement
+  - Import CTEs
+  - Logical CTEs
+  - Simple select statement
+
+### Example SQL with CTEs
+
+  ``` sql
+  with 
+
+  -- Import CTEs
+  suppliers as (
+      select * from {{ ref('stg_tpch__suppliers') }}
+  ),
+
+  nations as (
+      select * from {{ ref('stg_tpch__nations') }}
+  ),
+
+  regions as (
+      select * from {{ ref('stg_tpch__regions') }}
+  ),
+
+  -- Logical CTEs
+  locations as (
+      select
+          {{ dbt_utils.surrogate_key([
+              'regions.region_id',            
+              'nations.nation_id'
+          ]) }} as location_sk,
+
+          regions.region_id,
+          regions.name as region,
+          regions.comment as region_comment,
+
+          nations.nation_id,
+          nations.name as nation,
+          nations.comment as nation_comment
+        from regions
+      left join nations
+          on regions.region_id = nations.region_id
+  ),
+  
+  final as (
+      select
+          suppliers.supplier_id,
+          suppliers.supplier_name,
+          suppliers.supplier_address,
+          locations.nation,
+          locations.region,
+          suppliers.phone_number,
+          suppliers.account_balance
+      from suppliers
+      inner join locations
+          on suppliers.nation_id = locations.nation_id
+  )
+
+  -- Simple select statement
+  select * from final
+  ```

```

Comment:
I am not sure if we have talked about this before, but I am curious about this example CTE SQL. It seems like we are doing a lot of things we dont recommend here — aliasing columns in a marts model, bringing in columns we arent using later. Should we clean this up a bit? 

```suggestion
  ``` sql
  -- Jaffle shop went international!
  with 

  -- Import CTEs
  suppliers as (
      select * from {{ ref('stg_jaffle_shop__suppliers') }}
  ),

  nations as (
      select * from {{ ref('stg_jaffle_shop__nations') }}
  ),

  regions as (
      select * from {{ ref('stg_jaffle_shop__regions') }}
  ),

  -- Logical CTEs
  locations as (
      select
          {{ dbt_utils.surrogate_key([
              'regions.region_id',            
              'nations.nation_id'
          ]) }} as location_sk,

          regions.region_id,
          regions.region,
          regions.region_comment,

          nations.nation_id,
          nations.nation,
          nations.nation_comment
        from regions
      left join nations
          on regions.region_id = nations.region_id
  ),
  
  final as (
      select
          suppliers.supplier_id,
          suppliers.location_id,
          locations.region_id,
          locations.nation_id,
          
          suppliers.supplier_name,
          suppliers.supplier_address,
          suppliers.phone_number,
          locations.region,
          locations.region_comment,
          locations.nation,
          locations.nation_comment,
          
          suppliers.account_balance

      from suppliers
      inner join locations
          on suppliers.location_id = locations.location_sk
  )

  -- Simple select statement
  select * from final
  ```

```

Reply:
Thanks, this is all really great! I copied this from the template repo we have because I felt like we needed a little more than the example we had before, but could only find this as the shortest example of something that had more than one import & more than one logic step.

I appreciate the attention to detail 🙏 we absolutely should change those things - will get this in on Monday!

Reply:
Fixed [here](https://github.com/dbt-labs/corp/pull/63/commits/7151ce9c2c40374544f682f398e6ec82582c0076) with a minor change - I reorganized the import CTEs so it lists the data sets in the order they're used